### PR TITLE
Fixing some bugs with the AdaptiveOperationTracker

### PR DESCRIPTION
**Pull request by vgkholla**
1. Fixing a bug to handle a case where there can be no unexpired requests
2. Fixing a bug where code was written under the assumption that the tracker
cannot change value b/w the invocation of hasNext() and next(). This assumption
is not true if the router has multiple scaling units

Built and formatted

### CODE REVIEW

File: ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java [hunk 1]
```diff
@@ -123,14 +124,14 @@ public boolean hasNext() {
...
         // we are here because oldest request is past due
-        Map.Entry<ReplicaId, Long> oldestEntry = unexpiredRequestSendTimes.entrySet().iterator().next();
-        expiredRequestSendTimes.put(oldestEntry.getKey(), oldestEntry.getValue());
+        Map.Entry<ReplicaId, Pair<Boolean, Long>> oldestEntry = unexpiredRequestSendTimes.entrySet().iterator().next();
+        expiredRequestSendTimes.put(oldestEntry.getKey(), oldestEntry.getValue().getSecond());
         unexpiredRequestSendTimes.remove(oldestEntry.getKey());
         pastDueCounter.inc();
       }
-      unexpiredRequestSendTimes.put(lastReturned, time.milliseconds());
+      unexpiredRequestSendTimes.put(lastReturned, new Pair<>(false, time.milliseconds()));
```
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java]: Is it possible and if so, does it make more sense to do this (and incrementing `inFlightCount`) in `next()` and not in `remove()`. Seems like that is the right place to do it and you would not need to keep `lastReturned`.

**Comment by vgkholla** [comment]
@cgtz @pnarayanan Comments addressed. New commit is f0e4166

File: ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java (hunk 1 above)
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java]: The semantics for this iterator seems to be that "remove()" indicates that the `ReplicaId` will be used to send requests to - in order to respect that semantic, it has to be this way.

File: ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java [hunk 2]
```diff
@@ -148,10 +149,20 @@ public ReplicaId next() {
      * @throws IllegalStateException if no requests have been sent yet.
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java]: Removed the javadoc. Clarified this offline

**Pull request merged**